cmd/util/testutils: Add RunCmdAssertion test function

marclop opened the pull request:
## Description
<!--- Describe your changes in detail. -->
Adds a new function to the `testutils` package which allow easier
testing to be written on the `cmd` layer.

All the heavy lifting is done by `testutils.RunCmdAssertion` on the test
function, namely it:

    1. Initializes the `ecctl.App` singleton from a `testutils.MockCfg`.
    2. Sets the cobra `cmd.Command` Err and Out to `*Bytes.Buffer`s.
    3. Executes the command with the specified Args.
    4. Asserts the returned errors by `cmd.Execute()`
    5. Asserts the written data to both the `Err` and `Out` buffers.
    6. `ecctl.Cleanup` is called before returning so no instance is set.

It is important to note that global flags won't be parsed and instead,
must be provided as values to `testutils.MockCfg`. As new global flags
are added, the structure should be maintained to contain these.

Interaction with the `stdin` hasn't been written yet, but it will be a
follow-up patch.

Last, the tracker utils have been modified to use `DefaultTestFrequency`
when the host equals `api.DefaultMockHost` for ease of testing.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
Easier cmd layer testing!

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes.  Include -->
<!--- details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
`make unit`

## Types of Changes
<!--- What types of changes does your code introduce? Put an `x` in -->
<!--- all the boxes that apply: -->
- [x] Refactoring (improves code quality but has no user-facing effect)

marclop commented:
Commit breakdown:

* c5b20de: Adds the framework as described in the PR.
* 1386562: Adds tests for the `deployment create` command, with and without tracking.

karencfv commented on cmd/util/testutils/cmd.go:
```diff
@@ -0,0 +1,94 @@
+	args.Cmd.SetErr(cfg.Err)
+
+	if err := args.Cmd.Execute(); !assert.Equal(t, assertion.Err, err) {
+		t.Error(err)
+	}
+
+	if buf, ok := args.Cmd.OutOrStdout().(*bytes.Buffer); ok {
+		var got = buf.String()
+
+		// When the output contains the `--track` flag, removes the non-
+		// assertable duration time.
+		if strings.Contains(strings.Join(args.Args, " "), "--track") {
+			got = regexp.MustCompile(durationExpr).ReplaceAllString(
+				got, durationRepl,
+			)
+		}
+
+		if got != assertion.Stdout {
+			t.Errorf(`"Got stdout "%s" != want "%s"`, got, assertion.Stdout)
+		}
+	}
+
+	if buf, ok := args.Cmd.ErrOrStderr().(*bytes.Buffer); ok {
+		if got := buf.String(); got != assertion.Stderr {
+			t.Errorf(`"Got stderr "%s" != want "%s"`, got, assertion.Stderr)
+		}
+	}
```
I think I'd like this assertion code outside of the `TestCommand()` function, and in it's own function. What do you think?

marclop replied:
you mean as a private function?

karencfv replied:
yep :)

karencfv commented on cmd/util/testutils/cmd.go:
```diff
@@ -0,0 +1,94 @@
+
+// TestCommand tests a `*cobra.Command` and uses the testing.T struct to
+// return any unmatched assertions.
+func TestCommand(t *testing.T, args Args, assertion Assertion) {
```
I'm not sure I like `testutils.TestCommand` 🤔 How about `cmdutils.Test` or `testutils.AssertCommand` ? or something better that you might come up with 😄

marclop replied:
haha fair point, maybe `testutils.RunAssertion`, or just `Run`? `cmdutils.TestCommand` sounds good too.

karencfv replied:
I think just `Run` might be a bit confusing if the name of the package is `testutils`. I think I like `testutils.RunCmdAssertion` or `cmdutils.RunAssertion`. What do you think?

karencfv commented on cmd/util/testutils/testutils.go:
```diff
@@ -18,25 +18,78 @@
+}
+
+// MockApp initiates a mocked app from a MockCfg.
+func MockApp(t *testing.T, cfg MockCfg) func() {
```
This doesn't need to be exported any more, does it?

marclop replied:
Not really no, but it'd allow us to call it externally if we need? although I'm not sure we need to anymore

karencfv replied:
I don't think we need to any more 🤔 If we really need it we can export it later down the track. I also wouldn't want someone to find this function and skip the `TestCommand` function or whatever. What do you think?

marclop replied:
yeah Ok, I'll hide this function

marclop commented:
@karencfv I've addressed most of your comments, but left the two definitions.

I think it's better to have two deployment definitions with different values than 1, it doesn't mean that we should always use 2 different "provider" resource definitions but I did add them to the test to have different data. I've added the APM resource to the AWS definition.
